### ISSUE TITLE
[REVIEW]: dfba: Software for efficient simulation of dynamic flux-balance analysis models in Python

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2342]
**Submitting author:** @davidtourigny (<a href="http://orcid.org/0000-0002-3987-8078">David Tourigny</a>)
**Repository:** <a href="https://gitlab.com/davidtourigny/dynamic-fba" target ="_blank">https://gitlab.com/davidtourigny/dynamic-fba</a>
**Version:** v0.1.8
**Editor:** @marcosvital
**Reviewer:** @jdbrunner, @pstjohn, @synchon
**Archive:** <a href="https://doi.org/10.5281/zenodo.4009224" target="_blank">10.5281/zenodo.4009224</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/ba75c23c754d0dc44bf58eaf9f5629df/status.svg)](https://joss.theoj.org/papers/ba75c23c754d0dc44bf58eaf9f5629df)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/ba75c23c754d0dc44bf58eaf9f5629df"><img src="https://joss.theoj.org/papers/ba75c23c754d0dc44bf58eaf9f5629df/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/ba75c23c754d0dc44bf58eaf9f5629df/status.svg)](https://joss.theoj.org/papers/ba75c23c754d0dc44bf58eaf9f5629df)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://gitlab.com/davidtourigny/dynamic-fba" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@jdbrunner & @pstjohn & @synchon, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @marcosvital know.

✨ **Please try and complete your review in the next six weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/davidtourigny/dynamic-fba">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@davidtourigny) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/davidtourigny/dynamic-fba">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@davidtourigny) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://gitlab.com/davidtourigny/dynamic-fba">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@davidtourigny) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @jdbrunner, @pstjohn, @synchon it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1038/nbt1094-994 is OK
- 10.1038/nbt.1614 is OK
- 10.1016/S0006-3495(02)73903-9 is OK
- 10.1007/s00211-015-0760-3 is OK
- 10.1186/s12859-014-0409-8 is OK
- 10.1101/700112 is OK
- 10.1145/1089014.1089020 is OK
- 10.1186/1752-0509-7-74 is OK
- 10.21105/joss.00139 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02342/joss.02342/10.21105.joss.02342.pdf)

**Comment by marcosvital** [comment]
Dear @davidtourigny, @Midnighter and @carrascomj: your manuscript will be reviewed in this issue, and you can reply any comments and suggestions that the reviewers might address right here. Please take a last look at the manuscript proof above, since this is the version they will use to start the review.

**Comment by davidtourigny** [comment]
Thanks, the proof looks OK to me but welcome suggested improvements from the reviewers.

**Comment by marcosvital** [comment]
@jdbrunner, @pstjohn, @synchon, thank you all for for accepting review this submission for JOSS.
Even if you are not starting the review right now, please accept the invite, as it has an expiration date (there is a link under Reviewer instructions & questions and you should also get an email notification). Furthermore, please check the instructions and checklists above, and let me know if you need any assistance.

You can also tag  @davidtourigny, @Midnighter and @carrascomj if you need to ask specific questions about the submission.

**Comment by pstjohn** [comment]
This is a great contribution, and I think the paper is well-written. There's certainly a need for a high-performance, easy-to-use DFBA implementation, and I think this package goes a long way in solving that need. Writing something like this is not easy, it takes someone with knowledge / background in dynamic systems, software engineering (including writing compiled python extensions), and metabolic modeling to put it together, and David's done a great job with it.

I'll point out that I don't believe DFBAlab, the MATLAB package, actually reformulates the problem as a DAE. I think it instead just includes lexicographic constraints (for identifiability) and an LP-feasibility solution (to prevent LP crashes), but continues to solve the LP at each ODE integration. So the more advanced algorithm, combined with a C-level integrator, would likely mean that this version greatly improves on speed. DFBAlab, while widely used, is also not technically open-source.

I've submitted a few gitlab issues on better packaging / docker distribution, but otherwise my review is finished!

**Comment by marcosvital** [comment]
Thank you so much for you contributions, @pstjohn.

@synchon and @jdbrunner, I see that both of you have started the revision and left a few box unchecked: do you have any specific comments about those points to the authors?

 @davidtourigny, @Midnighter and @carrascomj, feel free to post here any changes you make on the manuscript and on the repository, ok?

**Comment by davidtourigny** [comment]
We have addressed @pstjohn 's issue on the repository https://gitlab.com/davidtourigny/dynamic-fba/-/issues/28

**Comment by synchon** [comment]
@marcosvital I have not completed the checks for those yet, will do it ASAP.

**Comment by synchon** [comment]
The package delivers what it aims to and in particular, I really like the level of abstraction used for the API design. Considering the complexity involved, it has been executed neatly.

I just have a few suggestions regarding the documentation and I have opened [https://gitlab.com/davidtourigny/dynamic-fba/-/issues/29](https://gitlab.com/davidtourigny/dynamic-fba/-/issues/29). Apart from this, I do not have any other suggestion and hence, I conclude my review.

**Comment by jdbrunner** [comment]
The package certainly fills a big need. For modeling interacting microbes, efficient dfba is a must, and I'm not aware of any current tools that do the job well. This package runs fast and is easy to use, with the only real drawback to simulating single models being the necessity of using docker. I am particularly impressed by how easy it is to control exchange flux lower bounds as functions of available resource. 

My only comment is that the  package as it currently stands does not allow for community simulations. A method for community dFBA in which microbes are coupled by a shared external metabolite pool would be a nice addition. With that suggestion, I conclude my review.

**Comment by davidtourigny** [comment]
Dear @synchon @pstjohn and @jdbrunner 

We (@Midnighter , @carrascomj and I) thank you very much for your positive reviews and great suggestions! We hope we have sufficiently addressed the issues you have raised on the repository.

Regarding @jdbrunner 's two last points: 1) distributing python wheels cross-platform for this project is work in progress (see issues [10](https://gitlab.com/davidtourigny/dynamic-fba/-/issues/10) and [26](https://gitlab.com/davidtourigny/dynamic-fba/-/issues/26) that unfortunately proves particularly challenging due to the compiled dependencies), but we hope at least the docker image proves sufficient for now, or that advanced users are able to follow the [instructions](https://dynamic-fba.readthedocs.io/en/latest/installation.html) for building from the source distribution; 2) a multi-species extension is certainly planned long-term for this project (see issue [14](https://gitlab.com/davidtourigny/dynamic-fba/-/issues/14)), but our primary objective was to implement a robust and easy-to-use package to make the most efficient DFBA algorithms available to users, before moving on to more complex models. We hope that this serves that purpose and look forward to feedback from the community regarding development of such extensions.

**Comment by marcosvital** [comment]
Dear @synchon @pstjohn and @jdbrunner

Thank you all very much for reviewing this submission. If you feel that all issues you have raised are sufficiently addressed, we will soon move on to publication.

**Comment by jdbrunner** [comment]
Yep, I'm satisfied!

**Comment by Midnighter** [comment]
At @marcosvital, one of the references which is cited as a pre-print was recently accepted. I would like to edit that reference before publication when I have the full information.

**Comment by davidtourigny** [comment]
@Midnighter I may be wrong, but there is nothing that prevents the paper being updated post-publication is there?

**Comment by marcosvital** [comment]
@davidtourigny, @Midnighter and @carrascomj, I'll check out if we can easily edit the paper after acceptance, and will let you know.

In the meantime, please take a final look at the proof (specially if any changes were made during the reviewing process), as we are close to finish here.

**Comment by marcosvital** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02342/joss.02342/10.21105.joss.02342.pdf)

**Comment by synchon** [comment]
Apologies for the delayed reply. I'm satisfied as well!

**Comment by davidtourigny** [comment]
Hi @marcosvital , the proof looks fine to me (no changes made during the review process). Thanks!

**Comment by marcosvital** [comment]
Hi again, @davidtourigny, @Midnighter and @carrascomj. Editing the paper after publishing is reserved for corrections, so it's probably better to wait if you are fine with it. @Midnighter, any news about the publication?

**Comment by davidtourigny** [comment]
I am happy to wait a couple more weeks if required (indeed, will be on vacation in any case).

**Comment by Midnighter** [comment]
I know that the DOI is going to be 10.15252/msb.20199235 but it's not available just yet. I think a few more days and it should be online. I'll edit the reference as soon as I know more.

**Comment by Midnighter** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #2342 with the following error: 

 Cloning into 'tmp/2342'...

**Comment by davidtourigny** [comment]
That is strange. It compiles fine with the preview service...

@marcosvital apart from this it appears we are now ready!

**Comment by davidtourigny** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02342/joss.02342/10.21105.joss.02342.pdf)

**Comment by davidtourigny** [comment]
@marcosvital all done!

**Comment by marcosvital** [comment]
@davidtourigny, @Midnighter and @carrascomj, we are almost done here. You will need to archive the last release of the package (on Zenodo, figshare, or other) - and if you already didn't do this, please let me know the package version number and archive DOI. Once this is done, we'll be ready to publish!

**Comment by davidtourigny** [comment]
@marcosvital great! We have tagged the latest release version 0.1.8 and archived it using Zendo DOI: [10.5281/zenodo.4009224](https://zenodo.org/record/4009224#.X0z_TUl7nOQ).

I don't seem to be able to add this information to the origional issue comment, sorry.

**Comment by marcosvital** [comment]
No need to worry about it, @davidtourigny, I'll do it.

**Comment by marcosvital** [comment]
@whedon set v0.1.8 as version

**Comment by whedon** [comment]
OK. v0.1.8 is the version.

**Comment by marcosvital** [comment]
@whedon set 10.5281/zenodo.4009224 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4009224" target="_blank">10.5281/zenodo.4009224</a> is the archive.

**Comment by marcosvital** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1689

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1689, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kthyng** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02342/joss.02342/10.21105.joss.02342.pdf)

**Comment by davidtourigny** [comment]
Everything looks good to me. Thanks

**Comment by kthyng** [comment]
A few tweaks for your paper (I'm not used to gitlab and looks harder to directly edit and submit a PR as an outside person than it is in github):

- [ ] "(as described in (Harwood, Höffner, & Barton, 2016))" — the reference shouldn't be parenthetical
- [ ] It's become the norm now in JOSS papers to make the "statement of need" a separate paragraph. You can keep all the words the same, but please change the statement of need part into it's own paragraph with it's own header.
- [ ] several of your references have the word "python" but it is not capitalized. Same for "c++11" the C and the subsequent python should be capitalized. You can preserve capitalization in bibtex by putting {} around the parts to preserve.
- [ ] Carefully look through the rest of the references to be sure they look exactly right.

**Comment by kthyng** [comment]
I see the version and Zenodo archive had been added and the metadata for the archive are correct, excellent!

**Comment by davidtourigny** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02342/joss.02342/10.21105.joss.02342.pdf)

**Comment by davidtourigny** [comment]
The latest version should take care of all your points @kthyng . Thank you

**Comment by kthyng** [comment]
@davidtourigny excellent looks good!!

**Comment by kthyng** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1038/nbt1094-994 is OK
- 10.1038/nbt.1614 is OK
- 10.1016/S0006-3495(02)73903-9 is OK
- 10.1007/s00211-015-0760-3 is OK
- 10.1186/s12859-014-0409-8 is OK
- 10.1145/1089014.1089020 is OK
- 10.1186/1752-0509-7-74 is OK
- 10.21105/joss.00139 is OK
- 10.15252/msb.20199235 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1300516339916984324) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1691
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02342](https://doi.org/10.21105/joss.02342)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kthyng** [comment]
Congrats to @davidtourigny on your new publication! Thanks to editor @marcosvital and to reviewers @jdbrunner, @pstjohn, and @synchon — this all wouldn't be possibly with your time and expertise.

(I'll close this issue once the doi resolves)

**Comment by davidtourigny** [comment]
Thank you so much @marcosvital @kthyng @pstjohn @synchon @jdbrunner for your time and hard work!

**Comment by Midnighter** [comment]
> Thank you so much @marcosvital @kthyng @pstjohn @synchon @jdbrunner for your time and hard work!

Thank you indeed! What a pleasant process it was overall. Definitely have definitely created a nice publishing experience.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02342/status.svg)](https://doi.org/10.21105/joss.02342)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02342">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02342/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02342/status.svg
   :target: https://doi.org/10.21105/joss.02342
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02342/status.svg)](https://doi.org/10.21105/joss.02342)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
